{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "对GATK DNAseq 分析流程进行总结，使用的gatk版本为GAT4.1.8.0\n",
    "\n",
    "数据下载：\n",
    "\n",
    "+ GATK介绍\n",
    "\n",
    "https://gatk.broadinstitute.org/hc/en-us/articles/360036194592-Getting-started-with-GATK4\n",
    "\n",
    "+ 下载链接\n",
    "\n",
    "https://github.com/broadinstitute/gatk/releases\n",
    "\n",
    "4.1.8： https://github.com/broadinstitute/gatk/releases/download/4.1.8.0/gatk-4.1.8.0.zip\n",
    "\n",
    "+ bwa下载链接\n",
    "\n",
    "https://sourceforge.net/projects/bio-bwa/files/bwa-0.7.17.tar.bz2/download\n",
    "\n",
    "https://sourceforge.net/projects/bio-bwa/files/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Germline SNV分析\n",
    "\n",
    "一般是多个样本进行germline分析，这样会更准确；\n",
    "\n",
    "也有单个样本的germline分析，需要使用机器学习算法；"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"./figures/fig1.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "分析流程主要包括：\n",
    "\n",
    "+ 数据预处理: BWA\n",
    "\n",
    "+ 标记重复: MarkDuplicates\n",
    "\n",
    "+ 碱基重新矫正：BaseRecalibrator/ApplyBQSR\n",
    "\n",
    "+ variantCalling: haplotyperCaller，由bam产生vcf/g.vcf文件\n",
    "\n",
    "+ JointCall: CombineGVCFs,将多个样本的g.vcf结合，产生vcf文件\n",
    "\n",
    "+ Genotyper： GenotypeGVCFs对上一步产生的vcf进行genotyping\n",
    "\n",
    "+ 使用VQSR进行矫正：VariantRecalibrator，ApplyVQSR\n",
    "\n",
    "+ 或者使用hard-filter进行过滤：variantFiltration\n",
    "\n",
    "参考文章：https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 数据预处理\n",
    "\n",
    "关于数据预处理，GATK官方也给出了The Best Practices:\n",
    "\n",
    "https://gatk.broadinstitute.org/hc/en-us/articles/360035535912\n",
    "\n",
    "+ bwa\n",
    "\n",
    "+ MarkDuplicates \n",
    "\n",
    "+ SortSam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variantion callings\n",
    "\n",
    "\n",
    "```\n",
    "$gatk HaplotypeCaller \\\n",
    "-R ~/gatk_files/human_g1k_v37.fasta \\\n",
    "-L ~/gatk_files/giab_agilent_sureselect_v5.list.bed \\\n",
    "-D ~/gatk_files/dbsnp_138.b37.vcf \\\n",
    "-I 151002_7001448_0359_AC7F6GANXX_Sample_HG003-EEogPU_v02-KIT-Av5_TCTTCACA_L008.posiSrt.markDup.bam  \\\n",
    "-O interval4180/hg003.g.vcf.gz \\\n",
    "-ERC GVCF \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## JointCall\n",
    "\n",
    "```\n",
    "$gatk CombineGVCFs \\\n",
    "-R ~/gatk_files/human_g1k_v37.fasta \\\n",
    "--variant hg002.g.vcf.gz \\\n",
    "--variant hg003.g.vcf.gz \\\n",
    "--variant hg004.g.vcf.gz \\\n",
    "-O corhort.g.vcf.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genotyper\n",
    "\n",
    "```\n",
    "$gatk GenotypeGVCFs \\\n",
    "-R ~/gatk_files/human_g1k_v37.fasta \\\n",
    "-V corhort.g.vcf.gz \\\n",
    "-O hg020304.vcf.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## VQSR \n",
    "\n",
    "### 可选步骤\n",
    "把SNP和indel分别取出来进行VQSR分析（也可以在一个文件中，但是要分两步做）\n",
    "\n",
    "```\n",
    "$gatk SelectVariants \\\n",
    "-R ~/gatk_files/human_g1k_v37.fasta \\\n",
    "-V hg020304.vcf.gz \\\n",
    "--select-type-to-include SNP \\\n",
    "-O hg020304_SNP.vcf.gz\n",
    "\n",
    "$gatk SelectVariants \\\n",
    "-R ~/gatk_files/human_g1k_v37.fasta \\\n",
    "-V hg020304.vcf.gz \\\n",
    "--select-type-to-include INDEL \\\n",
    "-O hg020304_INDEL.vcf.gz\n",
    "```\n",
    "\n",
    "\n",
    "### VQSR-snp\n",
    "\n",
    "```\n",
    "hapmap=~/gatk_files/hapmap_3.3.b37.vcf\n",
    " omni=~/gatk_files/1000G_omni2.5.b37.vcf\n",
    " kG=~/gatk_files/1000G_phase1.snps.high_confidence.b37.vcf\n",
    " dbsnp=~/gatk_files/dbsnp_138.b37.vcf\n",
    "\n",
    "#SNP\n",
    "$gatk VariantRecalibrator \\\n",
    "   -R  ~/gatk_files/human_g1k_v37.fasta \\\n",
    "   -V hg020304_SNP.vcf.gz \\\n",
    "   --resource:hapmap,known=false,training=true,truth=true,prior=15.0 $hapmap  \\\n",
    "   --resource:omni,known=false,training=true,truth=false,prior=12.0 $omni \\\n",
    "   --resource:1000G,known=false,training=true,truth=false,prior=10.0 $kG \\\n",
    "   --resource:dbsnp,known=true,training=false,truth=false,prior=2.0 $dbsnp \\\n",
    "   -an QD -an MQ -an MQRankSum -an ReadPosRankSum -an FS -an SOR \\\n",
    "   -mode SNP \\\n",
    "   -O output_SNP_vcf.recal \\\n",
    "   --tranches-file output_SNP.tranches\n",
    "```\n",
    "\n",
    "\n",
    "### ApplyVQSR-snp\n",
    "\n",
    "```\n",
    "$gatk ApplyVQSR -R ~/gatk_files/human_g1k_v37.fasta \\\n",
    "   -V hg020304_SNP.vcf.gz \\\n",
    "   -O ouput_SNP_recal.vcf.gz \\\n",
    "   --truth-sensitivity-filter-level 99.0 \\\n",
    "   --tranches-file output_SNP.tranches \\\n",
    "   --recal-file output_SNP_vcf.recal \\\n",
    "   -mode SNP\n",
    "```\n",
    "\n",
    "### VQSR-Indel\n",
    "\n",
    "```\n",
    "kgindel=~/gatk_files/1000G_phase1.indels.b37.vcf\n",
    " stdinde=~/gatk_files/Mills_and_1000G_gold_standard.indels.b37.vcf\n",
    " $gatk VariantRecalibrator \\\n",
    "   -R  ~/gatk_files/human_g1k_v37.fasta \\\n",
    "   -V ouput_INDEL_vcf.gz \\\n",
    "   --resource:mills,known=false,training=true,truth=true,prior=15.0 $stdinde  \\\n",
    "   -an QD -an MQ -an MQRankSum -an ReadPosRankSum -an FS -an SOR \\\n",
    "   -mode INDEL \\\n",
    "   -O output_INDEL_vcf.recal \\\n",
    "   --tranches-file output_INDEL.tranches\n",
    "```\n",
    "\n",
    "\n",
    "### ApplyVQSR-indel\n",
    "\n",
    "```\n",
    "$gatk ApplyVQSR -R ~/gatk_files/human_g1k_v37.fasta \\\n",
    "   -V hg020304_SNP.vcf.gz \\\n",
    "   -O ouput_indel_recal.vcf.gz \\\n",
    "   --truth-sensitivity-filter-level 99.0 \\\n",
    "   --tranches-file output_INDEL.tranches \\\n",
    "   --recal-file output_INDEL_vcf.recal \\\n",
    "   -mode indel\n",
    "```\n",
    "\n",
    "### 可选步骤 \n",
    "将分开的snp indel进行合并\n",
    "```\n",
    "$gatk MergeVcfs \\\n",
    "-I ouput_SNP_recal.vcf.gz \\\n",
    "-I output_INDEL_hard_filtered.vcf.gz \\\n",
    "-O hg020304_SNP_INDEL_filtered.vcf.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## hard-filter\n",
    "\n",
    "采用固定参数，对indel的结果进行硬过滤，可在--filter-name中添加字段\n",
    "\n",
    "```\n",
    "$gatk VariantFiltration \\\n",
    "-R ~/gatk_files/human_g1k_v37.fasta \\\n",
    "-V hg020304_INDEL.vcf.gz \\\n",
    "-O output_INDEL_hard_filtered.vcf.gz \\\n",
    "--filter-name \"QD2.0\" --filter-expression \"QD < 2.0\" \\\n",
    "--filter-name \"FS200.0\" --filter-expression \"FS > 200.0\" \\\n",
    "--filter-name \"ReadPosRankSum\" --filter-expression \"ReadPosRankSum < -20.0\" \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Single sample Germline SNV分析"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "参考文档：\n",
    "https://gatk.broadinstitute.org/hc/en-us/articles/360035535932-Germline-short-variant-discovery-SNPs-Indels-\n",
    "\n",
    "<img src=\"./figures/fig2.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Somtaic SNV分析\n",
    "\n",
    "需要normal文件，也可以tumor-only分析"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"./figures/fig3.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "分析流程主要包括：\n",
    "\n",
    "+ 数据预处理: BWA\n",
    "\n",
    "+ 标记重复: MarkDuplicates\n",
    "\n",
    "+ 碱基重新矫正：BaseRecalibrator/ApplyBQSR\n",
    "\n",
    "+ Mutect2，与germline不同，这里使用Mutect2进行variant calling\n",
    "\n",
    "+ calculate contamination: 给出一个cross-sample contamination，一些潜在的非somatic 突变\n",
    "\n",
    "+ variants 过滤：FilterMutectCalls对产生的vcf文件进行过滤\n",
    "\n",
    "+ 突变注释：Funcotator，对突变位点加上注释信息\n",
    "\n",
    "参考文档：https://gatk.broadinstitute.org/hc/en-us/articles/360035894731-Somatic-short-variant-discovery-SNVs-Indels-"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 数据预处理\n",
    "\n",
    "关于数据预处理，GATK官方也给出了The Best Practices:\n",
    "\n",
    "https://gatk.broadinstitute.org/hc/en-us/articles/360035535912\n",
    "\n",
    "+ bwa\n",
    "\n",
    "+ MarkDuplicates \n",
    "\n",
    "+ SortSam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Vairantion callings(Mutect2)\n",
    "\n",
    "+ tumor with matched normal\n",
    "\n",
    "参考文档： https://gatk.broadinstitute.org/hc/en-us/articles/360046788432-Mutect2\n",
    "\n",
    "```\n",
    "     gatk Mutect2 \\\n",
    "     -R reference.fa \\\n",
    "     -I tumor.bam \\\n",
    "     -I normal.bam \\\n",
    "     -normal normal_sample_name \\\n",
    "     --germline-resource af-only-gnomad.vcf.gz \\\n",
    "     --panel-of-normals pon.vcf.gz \\\n",
    "     -O somatic.vcf.gz\n",
    "```\n",
    "\n",
    "+ tumor-only \n",
    "\n",
    "```\n",
    "gatk Mutect2 \\\n",
    "   -R reference.fa \\\n",
    "   -I sample.bam \\\n",
    "   -O single_sample.vcf.gz\n",
    "```\n",
    "\n",
    "+ tumor-only with PoN(panel of normals)\n",
    "```\n",
    "  gatk Mutect2 \\\n",
    "  -R reference.fa \\\n",
    "  -I sample.bam \\\n",
    "  --germline-resource af-only-gnomad.vcf.gz \\\n",
    "  --panel-of-normals pon.vcf.gz \\\n",
    "  -O single_sample.vcf.gz\n",
    "```\n",
    "<p style=\"font-size:10px;color:blue;\"><b>af-only-gnomad.vcf.gz</b> 是funcotator中使用到的一个data source</p>\n",
    "\n",
    "+ Mitochondrial mode\n",
    "```\n",
    "gatk Mutect2 \\\n",
    "  -R reference.fa \\\n",
    "  -L chrM \\\n",
    "  --mitochondria-mode \\\n",
    "  -I mitochondria.bam \\\n",
    "  -O mitochondria.vcf.gz\n",
    "```\n",
    "\n",
    "+ Force-calling mode\n",
    "```\n",
    "gatk Mutect2 \\\n",
    "   -R reference.fa \\\n",
    "   -I sample.bam \\\n",
    "   -alleles force-call-alleles.vcf\n",
    "   -O single_sample.vcf.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estimate contamination reads\n",
    "\n",
    "包含两步\n",
    "\n",
    "### GetPileupSummaries\n",
    "需要两个输入文件\n",
    "+ germline variant sites VCF,比如gnomAD resource\n",
    "+ bam文件\n",
    "```\n",
    "gatk GetPileupSummaries \\\n",
    "   -I normal.bam \\\n",
    "   -V gnomad.vcf.gz \\\n",
    "   -L common_snps.interval_list \\\n",
    "   -O pileups.table\n",
    "```\n",
    "\n",
    "+ 产生的结果是这个样子\n",
    "```\n",
    "contig\tposition\tref_count\talt_count\tother_alt_count\tallele_frequency\n",
    " chr6\t29942512\t9\t0\t0\t0.063\n",
    " chr6\t29942517\t13\t1\t0\t0.062\n",
    " chr6\t29942525\t13\t7\t0\t0.063\n",
    " chr6\t29942547\t36\t0\t0\t0.077\n",
    "```\n",
    "\n",
    "\n",
    "### CalculateContamination\n",
    "\n",
    "计算the fraction of reads coming from cross-sample contamination\n",
    "\n",
    "+ tumor-only \n",
    "```\n",
    " gatk CalculateContamination \\\n",
    "   -I pileups.table \\\n",
    "   -O contamination.table\n",
    "```\n",
    "\n",
    "+ matched normal \n",
    "\n",
    "```\n",
    "gatk CalculateContamination \\\n",
    "   -I tumor-pileups.table \\\n",
    "   -matched normal-pileups.table \\\n",
    "   -O contamination.table\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter mutations\n",
    "\n",
    "对mutect2得到的vcf结果中的突变sites进行过滤\n",
    "\n",
    "```\n",
    " gatk FilterMutectCalls \\\n",
    "   -R reference.fasta \\\n",
    "   -V somatic.vcf.gz \\\n",
    "   --contamination-table contamination.table \\\n",
    "   --tumor-segmentation segments.tsv \\\n",
    "   -O filtered.vcf.gz\n",
    "```\n",
    "<p style=\"font-size:10px;color:blue;\"><b>segments.tsv, contamination.table</b> 均可由CalculateContamination计算得到</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 突变注释\n",
    "\n",
    "采用Funcotator对获得的突变位点进行注释\n",
    "\n",
    "参考文档：https://gatk.broadinstitute.org/hc/en-us/articles/360046786432-Funcotator\n",
    "\n",
    "使用方法看起来很简单，但是还需要进行学习，比如和annovar的区别，注释结果的结果。这个注释工具有个很大的resource data，30G~40G，需要下载解压使用。\n",
    "\n",
    "+ germline/somatic\n",
    "\n",
    "+ hg19/hg38\n",
    "\n",
    "```\n",
    "   ./gatk Funcotator \\\n",
    "   -R reference.fasta \\\n",
    "   -V input.vcf \\\n",
    "   -O outputFile \\\n",
    "   --output-file-format MAF \\\n",
    "   --data-sources-path dataSourcesFolder/ \\\n",
    "   --ref-version hg19\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
